### ISSUE TITLE
Odoo 9 requires connect permissions to 'postgres' database

### ISSUE THREAD

**Issue by LeartS** [odoo/odoo#12850]
**Impacted versions:**
9.0

**Steps to reproduce:**
In the postgresql pg_hba.conf file, don't allow the postgres user used by odoo to connect to the `postgres` database by removing the default lines and using a line like this in `pg_hba.conf`:

```
# TYPE  DATABASE            USER            ADDRESS                 METHOD
host    odoo-production     odoo-user       192.168.0.0/32            md5
```

**Current behavior:**
`psycopg2.OperationalError: FATAL:  no pg_hba.conf entry for host "<host>", user "<user>", database "postgres", SSL on`

This is because [the main always calls `_create_empty_database` if a `db_name` is set in the config](https://github.com/odoo/odoo/blob/9.0/openerp/cli/server.py#L131),
and [`_create_empty_database` tries to connect to the hardcoded `postgres` database](https://github.com/odoo/odoo/blob/7682760dcad232829fbb5d4221f66a1ebb9d6c91/openerp/service/db.py#L66).
In a deployment it's possible that odoo has access only to its own database and nothing else; so this will fail.

**Expected behavior:**
There should be a way to make Odoo work even without access to the `postgres` database.
Some alternatives:
* Ability to specify another "maintenance" database in the config file ([pgAdmin does this](http://dba.stackexchange.com/a/19154/56040))
* Ability to specify a config options to bypass [the check](https://github.com/odoo/odoo/blob/9.0/openerp/cli/server.py#L131) and assume that the db_name provided exists.
* Other possible config-based solutions that won't impact in any way current installations

cc. @xmo-odoo (I hope I pinged the right person)

**Comment by pedrobaeza** [comment]
@LeartS is this still happening?

**Comment by LeartS** [comment]
Yes, [`postgres` is still hardcoded in version 11](https://github.com/odoo/odoo/blob/b42b9c936b64dca72a41422e008537c34b691912/odoo/service/db.py#L91)

**Comment by pedrobaeza** [comment]
@Elkasitu are you the guy for checking this?

**Comment by Yenthe666** [comment]
Alternatively this is something for the deep framework guys (@odony or @xmo-odoo or @mart-e)

**Comment by mlaitinen** [comment]
This prevents us from using [Aiven DBaaS](https://aiven.io/postgresql) as our database. It's kind of possible but requires a lot of tweaking to make it work with connection pooling.

**Comment by LeartS** [comment]
Yes, as I've said in my PR, the missing ability to use a different database for maintenance creates problems in several uncommon but real situations:

* If access to `postgres` db is blocked because of over-zealous db admins whom you can't influence
* If there is some db connection middleware like pgbouncer that should do logic based on the database name. For example we have a pgbouncer instance that negotiates connections to two backend postgres instances: one instance of Postgres-XL for most of our project and one instance of standard PostgresSQL for Odoo, because postgres-XL does not support SAVEPOINT used by odoo. So we wanted to route client connections to the different instances based on the dbname: normal postgres if an odoo database, postgres-xl for all others db names. But as odoo also wants to connect to the postgres dbname, we cannot setup the route correctly in a single pgbouncer instance (unfortunately it doesn't support routing based on user) and have to bypass pgbouncer or set up another one on a different port
* If using different postgres "mods" like the mentioned Aiven
* Probably others..

**Comment by herawais** [comment]
Any update on this ?
I have a multi tenant Deployment of Odoo, almost 30 databases, it seems, all odoo instances (docker containers) creating connection to "postgres" database and the connection keeps idle or open.
hence it keeps eating more than 100 database connections

is there any workaround for that ? odoo should close the connections or it should not connect to "postgres" database because the --database filter is already applied in config file.

Thanks

**Comment by herawais** [comment]
> Yes, [`postgres` is still hardcoded in version 11](https://github.com/odoo/odoo/blob/b42b9c936b64dca72a41422e008537c34b691912/odoo/service/db.py#L91)

Today Odoo 16 released and still hardcoded in this file

**Comment by nar00n** [comment]
I am running Odoo Server 16.0+e-20230511 and the database is still hardcoded i guess, when i restrict access to the postgres database in the pg_hba.conf , odoo fails to start, while i have already defined my db name in the configuration

**Comment by HLFH** [comment]
The issue is still there on Odoo 17.

**Comment by kvaikla** [comment]
Why application deploy needs "Create DB" privilege at all? Or if it for some reason necessary, then it would be an option, do it or skip it.

**Comment by bendem** [comment]
> Why application deploy needs "Create DB" privilege at all?

Because Odoo expects to be able to work even if the DB doesn't exist yet, you point it at a PostgreSQL server and it will just work (tm), creating the database if it doesn't exist, also creating them when creating/restoring databases from the database manager.

You absolutely can remove the Create DB priviledge if you pre-create the database and the odoo user is owner of that DB. You won't be able to restore a dump through the database manager if you do.

Careful that Odoo will only recognise databases that it owns, databases that would be perfectly useable (with correct grants and all) won't be listed due to how Odoo discovers them.

---

I'll [repost the patch](https://github.com/odoo/odoo/pull/38926#issuecomment-1931505671) we have been using as a stop gap measure until Odoo takes this issue seriously, that patch hasn't been tested for multi-database setups:

```patch
*** odoo/sql_db.py      2024-02-06 10:00:07.608104282 +0100
--- sql_db.py   2024-02-06 10:17:37.479448663 +0100
***************
*** 787,796 ****
--- 787,799 ----
  def db_connect(to, allow_uri=False):
      global _Pool
      if _Pool is None:
          _Pool = ConnectionPool(int(tools.config['db_maxconn']))

+     if to == 'postgres':
+         to = tools.config['db_name']
+
      db, info = connection_info_for(to)
      if not allow_uri and db != to:
          raise ValueError('URI connections not allowed')
      return Connection(_Pool, db, info)
```

**Comment by ml31415** [comment]
> Because Odoo expects to be able to work even if the DB doesn't exist yet, you point it at a PostgreSQL server and it will just work

Making the assumption, that you have database create rights is imho exactly as wrong, as assuming that the database is just there. From any mature application, I'd expect to see what you have and run, if possible. Clearly, it's possible to run odoo without access to `postgres`, so why depend on it? Any database backup features etc can fail separately. But this should not be a blocker for vanilla installations.

**Comment by ml31415** [comment]
@Yenthe666 as you already added the first labels to this issue: Would you mind adding labels for odoo 17? And maybe also editing the name? This is unfortunately still an unsolved issue.

**Comment by Yenthe666** [comment]
Done :-)
